[NEEO] Initial Contribution of NEEO binding/transport #2350

tmrobert8 opened the pull request:
Signed-off-by: Tim Roberts <troberts@bigfoot.com> (github: tmrobert8)

Initial contribution of NEEO binding and NEEO transport to fully support the NEEO brain/remote

**Binding Capabilities**
The binding will discover the NEEO brains and provide you with a channel for each recipe - allowing you to start and stop NEEO recipes from openHAB.

**Transport Capabilities**
The more powerful transport does MUCH more. The transport will allow you to expose openHAB things/channels to the NEEO brain - allowing you to control openHAB via the remote control. Example: I can expose my music system (Russound binding in openHAB) as an audio device on the NEEO and can turn it on/off with the remote power buttons, change the volume with the volume buttons, switch sources with the remote display and even display the album cover art on the remote display. VERY EXCITING.

tmrobert8 commented:
Latest binding/transport:
[org.openhab.binding.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1677673/org.openhab.binding.neeo-2.3.0-SNAPSHOT.zip)
[org.openhab.io.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1677674/org.openhab.io.neeo-2.3.0-SNAPSHOT.zip)

tmrobert8 commented on addons/binding/org.openhab.binding.neeo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,161 @@
+	            <default>true</default>   
+	        </parameter>
+	        
+	        <parameter name="localIpAddress" type="text" required="false" >
```
This was put together before the PR (forget the number) that added setting of the primary address - until that point in time, the network service/netutil was always returning the wrong ip address on my machine and I needed that to work correctly - I'll remove it now that we can do that via the platform

tmrobert8 commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoDeviceHandler.java:
```diff
@@ -0,0 +1,288 @@
+
+            final ThingBuilder thingBuilder = editThing();
+            thingBuilder.withChannels(channels);
+            thingBuilder.withLabel(thing.getLabel());// bug to save it
```
I'll need to retest but back in OH2.0 - there was a bug where the label was being lost when you did an updatething.  Probably has been fixed by now - but I'll retest and get back to you on this

tmrobert8 commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoBrainHandler.java:
```diff
@@ -0,0 +1,408 @@
+            }
+        }
+
+        // finally - fallback to what openhab says
```
Again - predated that PR setting PRIMARY_ADDRESS and I needed a way to discover the local address to listen on - I'll update this (and the transport) to reflect being able to use the PRIMARY_ADDRESS (and add information to the README.MD as well)

tmrobert8 commented:
@triller-telekom 
To address your general concerns:
1.  Modeling - trust me - you don't want me to create a graphic (I can't draw a stick figure!) :)
The modeling reflects (almost exactly) how the brain itself has things modelled - I simply followed it along because it's easier to make changes as they make changes (that I need to reflect).  But here is the overview of it:
Brain (bridge) - contains information about the brain itself and is the source of forward events.  A brain contains zero to many rooms.
Room (bridge) - contains information about the room and is where scenarios/recipes (both device based *and custom*) are contained/executed.  A room can contain zero to many devices.
Device (thing) - contains information about the device and contains executable macros. 

Those using the brain need the ability to execute scenarios/recipes (which are based on the room) and the ability to execute macros (which are based on the device).  

Note: both the room/device actually contain alot more - but none of it is accessible from the API yet.  I can get the information - but do nothing with it (yet).  I'm hopeful one day to add more stuff to both levels.

2.  TRIGGER - there apparently is alot of confusion on when to use these. Originally these were state channels but of the prior THREE reviewers that commented in the other PR - two wanted these channels to be TRIGGERS because they are less like state but more like events.  I'm really on the fence (I can see both arguments being valid) and went with what they wanted.  The JSON array was because it's easier to parse in rules apparently than a comma-delimited state - again, didn't have a strong opinion one way or the other and went with it.

3.  Threading - absolutely need threading.  The initialization task is threaded because I'd get blacklisted otherwise.  Because it's threaded - you can reenter the initialization on any number of scenarios (config update being the most likely) and I have to handle this.  Note: this was also written back before I posted how initialize() can be reentered by multiple threads via the platform itself (one of which was solved via https://github.com/eclipse/smarthome/issues/3351).

4.  Lastly - magic numbers.  I'm not a big fan of constants for things that are specified in only once.  To me, you add unnecessary complications (more code to maintain, more to document, harder to understand since you now reference a contants class and easier to screw up).  I'll change those if you feel strongly the opposite way - let me know.

tmrobert8 commented:
@triller-telekom 
Made the changes and committed.  Just for you information, I will be leaving on vacation until Feb 10th (as of Feb 2nd) and won't be responding to anything during those dates - so you have time to look through the rest of the binding/transport (unless of course you want to finish it all off in the next two days - then you have very little time!).    :)

triller-telekom commented:
> Threading - absolutely need threading. The initialization task is threaded because I'd get blacklisted otherwise. Because it's threaded - you can reenter the initialization on any number of scenarios (config update being the most likely) and I have to handle this. Note: this was also written back before I posted how initialize() can be reentered by multiple threads via the platform itself (one of which was solved via eclipse/smarthome#3351).

Just one quick feedback here before your vacation. Its fine if you initialization task needs longer that you schedule it in a different thread. However you should not run into the situation that the initialization is called multiple times. If a `Handler` receives a new `Configuration`, `thingUpdated` is called which by default disposes the `Handler`. At least that is what the `BaseThingHandler` is doing. You can overwrite both methods to clean up your threads and then get rid of all your `AtomicReference`s.

tmrobert8 commented:
@triller-telekom 
I hope you don't mind - I love discussing this type of stuff.  Both @martinvw and I had some good discussion on my use of 'final' for local variables.

To your point - the dispose only marks the thread as interrupted and it actually won't be interrupted until checkInterrupted is called - so you'll still have multiple initializes running simultaneously depending on how the JVM scheduler handles things.

"you should not run into the situation" - that's the other point I'd like to make.  I did run into that situation due to a design error (which brought up that original forum message) and I lost 3 days of debugging figuring it out on the Russound binding.  By doing what I'm doing, I'm protecting myself and the users of this binding against errors in the underlying platform as well.

Lastly, this isn't strictly about double initialization either, since the Brain instances and BrainApi instance is shared across the BrianHandler/RoomHandler/DeviceHandler (via getBrainApi()/getBrain() calls) - I get other interesting multithreading access scenarios.  A forward actions from the servlet and a configuration update at the same time, a device channel update (probably from the transport) and a configuration update at the same time, etc.  

openHAB is a multiple threading environment and BridgeHandlers inherently must deal with multiple thread (since they can be referenced by child handlers which may be called from separate threads). I also know for a fact that I can get calls to handleCommand during a config update as well (even though that doesn't apply to this particular handler) - demonstrating that a regular thing handler is still not safe from multiple thread interactions.

Now - given that and apparently your dislike for AtomicReferences - I could make them all basic references protected by one overall ReentrantLock (or maybe a readerwriterlock makes more sense).  In fact, it makes more sense that way for this handler (and will clean up another multithread issue I just realized).  What do you think?

tmrobert8 commented:
New version of transport:

[org.openhab.io.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1686818/org.openhab.io.neeo-2.3.0-SNAPSHOT.zip)

Changes:
1.  Added ability to show the EUI from the brain page
2.  Added ability to display the log file from the brain to the brain page

tmrobert8 commented:
@triller-telekom 
FYI - back from vacation now (although swamped with everything pending).  Let me know if you've gotten any further with the review

triller-telekom commented:
@tmrobert8 Sorry I just found some time to have a look at your binding again. While looking at the checked out code I noticed that it is a bit cluttered by `Objects.requireNonNull()` calls. You are doing these checks for almost all of your methods even though you are just calling them by yourself.

For exactly this use-case we have introduced `null` annotations, see https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style point 12. Please annotate your classes with them and get rid of those checks, it makes the code much more readable.

Regarding the locking: if you do find problems within the [ESH core](https://github.com/eclipse/smarthome) please feel free to file bugs or provide PRs to fix them so all users of the framework will benefit from it. We all know that it is hard to write multi-threaded code but we should not make it more complicated by synchronizing on all variables/methods... If you have to share the `brain` instance over all handlers make sure you synchronize the necessary methods within this object and please remove the `AtomicReferences`.

Regarding your modeling of things, please give me some more time to investigate it.

tmrobert8 commented:
@triller-telekom 
Thanks for taking a look again.  I'd love to implement the null annotations but that's a HUGE job/change to the whole structure (I'm doing that in the sony binding and am finding all sorts of 'interesting' things with the null annotations).  Since this is such a large binding and the firmware changes often - I've been having trouble getting this being initially merged (as shown by you being the forth reviewer now).  The next firmware is probably within a few weeks (and may or may not cause more changes) - I'm really trying to focus this on critical stuff to get it merged - then do the 'nice to have' stuff in a future with much smaller, easier to review PRs.  So if you don't mind, can I ignore the null annotation suggestion for now?

Note: it **appears** that you can't simply drop the null checks and blindly replaces them with null annotations, in the sony binding where I am converting it to null annotations - there are definitely areas where a null can still get in regardless of the annotation (mainly where you have interactions with other libraries that don't support null annotations [gson]).  So it's not a simple conversion process (plus there are weird coding things [especially around collections] that make the code a bit 'loopy' to implement null annotation correctly).

If you look at the latest code, I've eliminated almost all AtomicReferences.  I prefer AtomicReferences over synchronization in almost ever case for numerous reasons (in a nutshell, atomics are more concise, easier to read/maintain, generally faster [grain of salt on that statement however] and are less prone to programmer errors).

If we can - can we just concentrate on anything critical.  The release cycle of NEEO is suppose to pick up and will make this harder and harder to get initially merged.  My goal is to get the initial codebase merged and then have smaller, easier to manage, PRs in the future - I just need to get this initially merged to make this whole process easier...

triller-telekom commented:
For sure null annotations will not defend you from being called with `null` at runtime. nevertheless most methods are only called by yourself... Also some comments of your `Objects.requireNonNull()` are misleading due to copy&paste, see https://github.com/openhab/openhab2-addons/pull/3191/files#diff-a0dd9523286be7330f84fc3f40bb65b4R75

Another things I just found is: Why does you `device thing` contain a configuration parameter `roomKey`? You intentionally created the `room bridge` and attach the `device thing` to it. So in any case you can always ask the bridge of your things which `roomKey` it has. Or am I missing anything there?

I am sorry that Neeo gives you a hard time keeping up with their API, but that should only affect the communication with them not the modelling for OH, right?

tmrobert8 commented:
Please point out anywhere I have a copy/paste error (which is not surprising in the least).  

Good catch on the roomKey - I'll be changing that shortly.  

If you could point to an NEEO API, I'd love to use it  ;)   Seriously - the binding is using URLs that I've wiresharked off their app.  The transport uses something that mirrors (in part) their node.js SDK internals.  So the point is - they change their firmware and their own app/SDK - I then need to make the same changes in the binding and transport SDK.  There is no public API.  Most of the time, there are no changes or simple enhancements.  However, the firmware between you and @martinvw review did happen to have a large change in it (which changed ALOT of the transport internals).  I will agree that the binding has been fairly stable in it's model - the transport is where most of the changes occur however.

tmrobert8 commented:
"nevertheless most methods are only called by yourself" - lol don't get me started on that statement.  I'm a designer at heart and love to argue stuff and that's one heck of a statement you just made - but I'll leave that for a different time  ;)

tmrobert8 commented:
@triller-telekom 

Just committed the changes from the latest:
1.  Removed roomKey from deviceconfig and related classes
2.  Scanned the project for requireNotNull and requireNotEmpty and fixed about 8-10 copy/paste errors (scanned both transport and binding code).  Should be correct now but of course, let me know if it isn't.

triller-telekom commented:
@tmrobert8 Thank you for your changes, but next time, PLEASE create a new commit. by amending your previous commit you make it hard for me to see the differences, especially in a PR with >18k lines of code.

I'll let you know once I have more feedback for you.

tmrobert8 commented:
Lol every reviewer is different - I’ve had a few that absolutely didn’t want that to happen (polluted the tree with extra nodes).  I’ll do that for future changes for you

Sent from my iPhone

> On Feb 15, 2018, at 3:58 AM, Stefan Triller <notifications@github.com> wrote:
> 
> @tmrobert8 Thank you for your changes, but next time, PLEASE create a new commit. by amending your previous commit you make it hard for me to see the differences, especially in a PR with >18k lines of code.
> 
> I'll let you know once I have more feedback for you.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/NeeoDeviceProtocol.java:
```diff
@@ -0,0 +1,125 @@
+                    api.triggerMacro(neeoRoom.getKey(), neeoDevice.getKey(), macroKey);
+                    callback.scheduleTask(() -> {
+                        refreshMacroStatus(macroKey);
+                    }, 500);
```
What if the device takes longer than 500ms to change its status based on the sent command? Will you miss that the device has accepted/not accepted the desired state change?

If the brain doesn't provide any push mechanism, wouldn't it make sense to have a polling job running that polls the status every x seconds and set the polled state in OH2? The hue binding does that for example, because the bridge also doesn't have a push mechanism.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/NeeoDeviceProtocol.java:
```diff
@@ -0,0 +1,125 @@
+                }
+            } catch (IOException e) {
+                logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
+                // callback.statusChanged(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```
if you cannot communicate with your device, why not setting it to `OFFLINE`?

tmrobert8 replied:
As you can see - originally I did.  However, realize we just got to firmware .43 and there are still issues on their side that cause random errors.  Because of that, I don't want to mark it offline when we just hit a bug in their firmware.  When we get to an official 1.0 release - let's revisit this.  If you'd like, I can either leave it as is or maybe document that?

triller-telekom replied:
Please document it somewhere, so you do not forget to remove this hack.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoDeviceHandler.java:
```diff
@@ -0,0 +1,304 @@
+        final String groupId = channelUID.getGroupId();
+        final String channelId = channelUID.getIdWithoutGroup();
+
+        ((StatefulHandlerCallback) protocol.getCallback()).removeState(channelUID.getId());
```
Why are you caching the old state? You say that you only want to communicate if it has changed, but why is that? is the communication with the device really that expensive in terms of it takes **very** long or consumes a lot of computation power on the brain?

tmrobert8 commented on addons/io/org.openhab.io.neeo/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,61 @@
+ org.osgi.util.tracker,
+ org.slf4j
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.io.neeo
```
Think you are looking at the wrong manifest - this one is the IO one which doesn't have a handler.  The binding manifest does export handler

tmrobert8 commented:
@triller-telekom 
Just committed the changes except for the two above - you brought up some good points btw!

"I guess you are also familiar with the OSGi framework?" - that's not what I meant by that statement.  The standards you hold methods to should be consistent regardless of the visibility because that visibility can easily change.  In fact, with today's IDEs - changing the visibility of a different class (that isn't opened) is a simple refactoring (literally two actions).  If you don't hold private methods accountable like public methods - you will start introducing errors because someone will refactor without realizing the implications.  I could go on an on about this - but this isn't the place.

As for null annotations, while I see them as valuable for clearing potential issues while coding (due to the static analyzing of source files), I do not see why you'd want to eliminate runtime checks in favor of them - that would make debugging a much MUCH harder task (especially if you get an NPE somewhere in a stream that could have multitudes of possibilities on what went null) - it if was a parm, I rather know up front something was null at runtime rather than having to deduce it from log files.  The readibility cost is very minor compared to what is gained at runtime.

ppieczul commented:
Hi @tmrobert8.
I cleaned up my local network DHCP assignments and noticed a weird behavior with the transport. I basically changed the IP address of the brain and it acquired it after the lease expired. The transport stopped functioning at that moment. I tried to reproduce it with debug logs enabled - setup the transport and brain so they were working. I then changed the IP address on the DHCP again and restarted the brain, so it acquired the new IP. Then I uninstalled the transport bundle from the console and installed it again - still not working. Here are some logs from when transport lost the connection to the brain. It looks like the old IP address was still in use and the transport indefinitely was throwing exceptions.
I used the latest transport jar build from the artifactory from this PR.
Thanks
Pawel
[neeo.log](https://github.com/openhab/openhab2-addons/files/1745872/neeo.log)

tmrobert8 commented:
@ppieczul 
Yeah - I've known about this one for awhile and it's really something I can't solve.  I believe the issue is in the MDNS implementation on the brain (I posted this on planet about a year ago - but their sole developer [at the time] never got around to answering me - I'll repost since they have a team now and see if it goes anywhere).  Basically what happens is that the brain never sends out it's updated IP address and because of that, MDNS never get's updated and therefore the transport is never able to update to the new address.  Note: there *may* be a secondary issue in that there is a bug in JMDNS (which openHAB uses that doesn't seem to requery MDNS also and uses a cached version) - I've never been able to track it down to open an issue however - so that's more of a 'I think' statement.

Here's a simple test case you can try:
1.  Assign a new IP address and let the brain acquire the new IP address
2.  From the command line, type "dns-sd -B _neeo._tcp" - you'll see the old IP address of the brain still listed.

The brain does eventually update MDNS and the transport will eventually reconnect to it - but it doesn't happen immediately.  I'll probably update the README to strongly recommend assigning an IP address to the brain to avoid this and other issues related to IP address changes

tmrobert8 commented:
@triller-telekom 
Thank you for your time in reviewing this.  At this time, we may as well stop the review however.  If you are insisting on the nullable tags (which is fine) and based on the Sony addon (which is roughly the same size), looks like we are talking about 6-8 months before I can get that done (I'm on month 3 on the Sony one and only 75% done) - I simply don't have alot of time to address something like this.

I would like to challenge the guideline however (I like the tag but you absolutely need the runtime check as well) but am not sure what the proper forum is for that (maybe @kaikreuzer would like to chime in on that).

As for the marketplace, I can't publish this because according to @kaikreuzer - IO (ie the transport) is not supported by the marketplace and it's really the critical part of the process.

Finally, I disagree with the HUE design pattern for discovery.  I originally duplicated that but found it too complicated, not easily understood, inappropriate for the scope of a thing, and complicates the class dependency graph too much.  Discovery absolutely needs to deal with the ThingRegistry if you want a seamless user experience.  (Too) simple example would be the device announcing itself on a different IP address or port number (darn Sony tvs).  You don't want new discovery results but to have the existing thing's configuration updated instead.  The Thing can't do this on it's own because (properly) it never is exposed to the new announcement - the only place (logically) for this to happen within the discovery scope and to do that you need access to the ThingRegistry.  Because of that, you can simplify the design like I did and ALL of the discovery occurs in the scope of a discovery class (not spread across the thing handler factory, thing and discovery classes like the hue pattern does).  I could go on but again - this isn't the right place for it.

Again - thanks for the time you put on this.  Unfortunately I simply don't have the time to update to the Nullable pattern (plus I really think that pattern needs to be tweaked in the guidelines still before it should be applied across everything)

triller-telekom commented:
@tmrobert8 Thank you for you reply.

As I read from it you are not happy about our nullable guidelines. May I ask you to have a look on what we discussed before introducing them here: https://github.com/eclipse/smarthome/issues/3624 and afterwards add your additional thoughts here https://github.com/eclipse/smarthome/issues/4321, so they are not lost? We really appreciate improvements of this kind!

tmrobert8 commented:
@triller-telekom @kaikreuzer 
Just committed a new version (I'll fix those pom conflicts in the next commit).  I added the null annotations through out.  However, to @triller-telekom point - this will change frequently since I keep the internal SDK in sync with NEEO's official SDK and will keep changing/adding features as needed.

So the question becomes - do we want to have this added to the repository?  

On the one side, it would be nice to get it in because whichever version will work with the remote just fine (they'll just be missing the latest features) and with the remote being such a GREAT front end to openhab - people will likely be using it (already have redirected a bunch of people over on the NEEO forums to the openHAB forum for installation directions).  Would be much less work if they could just install openHAB and install them directly (once you start talking directories - many non-techs roll there eyes thinking it's for professionals - that's been my biggest issue on that forum)

On the other hand, people may just get use to going to the thread to download the latest version anyway and with the amount of changes that will always occur - it may be too much hassle.

I don't know the answer to this - maybe a middle ground where we have it in openHAB and have quarterly 'official' releases (and the forum for interim one).  Just thinking out loud.

At any rate, I'd like to hear input on this

tmrobert8 commented:
Latest files:
[org.openhab.binding.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1876895/org.openhab.binding.neeo-2.3.0-SNAPSHOT.zip)
[org.openhab.io.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1876896/org.openhab.io.neeo-2.3.0-SNAPSHOT.zip)

tmrobert8 commented:
Fixed files:
[org.openhab.binding.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1879927/org.openhab.binding.neeo-2.3.0-SNAPSHOT.zip)
[org.openhab.io.neeo-2.3.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1879928/org.openhab.io.neeo-2.3.0-SNAPSHOT.zip)

triller-telekom commented:
@tmrobert8: What is the status of this binding on your side? You said that the API is not very stable and changes often so you have to adapt it rather quickly and this isn't really suitable for a couple of month old openHAB installation, right?

How about just publishing your binding to the marketplace for people to download it? This way they can always use your newest version, even if they run on an older openHAB stable version (assuming you dont make use of any new ESH/openHAB features). The only problem with that would be that the marketplace only supports "binding bundles". So you would have to move your "io" code into your binding bundle and publish this as one.

Would that be an option?

Also if the API stabilizes, we could think of integrating your binding into the official repository. The question is what do we do with this PR? Do you want to continue working on the binding on this PR for everyone to follow? In this case we can leave it open, but please mark it with "WIP" so it is obvious that you are still working on it and its not a binding waiting for a final review.

tmrobert8 commented:
Support for 50.x:
[org.openhab.binding.neeo-2.3.0-20180515-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/2004771/org.openhab.binding.neeo-2.3.0-20180515-SNAPSHOT.zip)
[org.openhab.io.neeo-2.3.0-20180515-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/2004775/org.openhab.io.neeo-2.3.0-20180515-SNAPSHOT.zip)

tmrobert8 commented:
@triller-telekom 
I haven't responded because I wanted to see what the volume of changes are happening in the firmwares.  Since you've posted, there has been around 15 non-public firmware releases and 2 public (with one forthcoming) and there hasn't been any breaking changes in any of them (certainly some additions - but nothing breaking).

I know this binding is being used by quite a few people and has been fairly stable at this point and I feel better continuing the merge process at this point.  As to the marketplace, originally it was part of it but Kai felt strongly that the 'io' portion needed to be in the transports area (since that's what it really is) and I took it off the marketplace at that point.

tmrobert8 commented:
Test for DiLucca

[org.openhab.io.neeo-2.4.0-20180706SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/2170539/org.openhab.io.neeo-2.4.0-20180706SNAPSHOT.zip)

tmrobert8 commented:
Support for 51.1 (rename from .zip to .jar - please note these say 2.4 but will work in earlier versions):
[org.openhab.binding.neeo-2.4.0-20180711-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/2185766/org.openhab.binding.neeo-2.4.0-20180711-SNAPSHOT.zip)
[org.openhab.io.neeo-2.4.0-20180711-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/2185767/org.openhab.io.neeo-2.4.0-20180711-SNAPSHOT.zip)

Binding changes:
None

Transport changes:
1. Implemented new LIST type
2. Updated to new NEEO SDK naming standards
3. Updated to new notification API for 51.1+ firmware
4. Added firmware to brain status page
5. Fixed numerous bugs on things page
6. Fixed search to match tokensearch exactly

Strongly recommended to drop and readd your existing devices

tmrobert8 commented:
@triller-telekom @martinvw 
Believe we have a windows of a month or two if you want to try to get this reviewed again

Tim

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/AbstractThingHandler.java:
```diff
@@ -0,0 +1,69 @@
+     * @return the {@link NeeoBrainApi} or null if not found
+     */
+    @Nullable
+    public NeeoBrainApi getNeeoBrainApi() {
```
Instead of having this method and the one below twice (copy&paste), how about moving it to your util class with a `Bridge bridge` parameter?

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoBrainHandler.java:
```diff
@@ -0,0 +1,389 @@
+import javax.servlet.ServletException;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNull;
```
Please remove references to this `NonNull` throughout the PR.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoBrainHandler.java:
```diff
@@ -0,0 +1,389 @@
+            updateStatus(ThingStatus.ONLINE);
+            NeeoUtil.checkInterrupt();
+            if (config.getCheckStatusInterval() > 0) {
+                NeeoUtil.cancel(checkStatus.getAndSet(scheduler.scheduleWithFixedDelay(() -> {
```
Why do you cancel this task immediately?

tmrobert8 replied:
I cancel the prior task if it's been scheduled (I get the existing task, then set the new task [atomically with the get] and then cancel the existing task if there was one).

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoBrainHandler.java:
```diff
@@ -0,0 +1,389 @@
+                    } catch (IOException e) {
+                        logger.debug("IOException occurred deregistering the forward actions: {}", e.getMessage(), e);
+                    }
+
```
blank line, please remove

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoHandlerFactory.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * Unsets the {@link HttpService}
```
A lot of javadocs in this class ave copy&paste errors, please fix.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoHandlerFactory.java:
```diff
@@ -0,0 +1,171 @@
+
+        if (localRegistry != null) {
+            scheduler.schedule(() -> {
+                localRegistry.startScan(NeeoConstants.BINDING_ID, null);
```
please remove the registry and replace it with you own `DiscoveryService` which you can start here, similar to https://github.com/eclipse/smarthome/blob/bc9d6284c54ca38b13436abf733d567b1e3c5d6d/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java#L104-L109

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoHandlerFactory.java:
```diff
@@ -0,0 +1,171 @@
+     * @param discoveryServiceRegistry the non-null {@link NetworkAddressService} to use
+     */
+    @Reference
+    protected void setDiscoveryServiceRegistry(DiscoveryServiceRegistry discoveryServiceRegistry) {
```
A binding should not interact with the `DiscoveryServiceRegistry`, please remove

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/package-info.java:
```diff
@@ -0,0 +1,14 @@
+/**
```
We are not using these package info files, please remove.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/NeeoRoomDiscoveryService.java:
```diff
@@ -0,0 +1,237 @@
+
+    /** The thing registry that we use */
+    @NonNullByDefault({})
+    private ThingRegistry thingRegistry;
```
A discovery service must not use the thingregistry, please remove.

tmrobert8 replied:
Let's table this and the other discovery comments to see what the end result of https://github.com/eclipse/smarthome/issues/5582 (which I'll be adding comments to eventually).  I really don't like the approach you described because it sprinkles the discovery around a few classes.

triller-telekom replied:
Alright, and while you are positing to eclipse/smarthome#5582, please link this conversation.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/NeeoRoomDiscoveryService.java:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        logger.debug("Scanning all things for NeeoBrainHandlers");
+        for (final Thing thing : localThingRegistry.getAll()) {
```
Since you do not have the thingregistry anymore you can create a discovery service from the handler factory and pass the brain bridge to it so it can access its api and query the hardware directly.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/package-info.java:
```diff
@@ -0,0 +1,14 @@
+/**
```
Please remove all these `package-info.java` files.

tmrobert8 commented on addons/io/org.openhab.io.neeo/README.md:
```diff
@@ -0,0 +1,335 @@
+
+## openHAB Primary Address
+
+This transport will use the primary address defined in openHAB to register itself as an SDK with the NEEO Brain (allowing the NEEO Brain to discover your things and to callback to openHAB to retrieve item values).  If you have multiple network interfaces on the machine that openHAB runs on or if forward actions are not being recieved, you'll likely need to set the primary address configuration fi ...
```
Yes - I'll be updating this for the new interface shortly - thanks for pointing out the doc note however (forgot about that).  I'm assuming your note about the line break applies to ALL sentences within the readme(s)?

tmrobert8 commented on addons/io/org.openhab.io.neeo/README.md:
```diff
@@ -0,0 +1,335 @@
+There are two types of devices listed here:
+
+1.  openHAB things - any OH2 thing in the openHAB system and optionally (see configuration), any NEEO binding thing 
+2.  virtual things - these are things you defined that can mix/match any OH1 or OH2 item
```
Unless I'm missing what you are saying - both those sentences are correct (maybe I need to add some more wording to make it clear).  

Basically - things are exposed to NEEO as a device and the items within the thing are exposed as NEEO capabilities.   You need both to work.

You can expose two different types: 
1. a native OH2 thing - the thing is exposed directly as a device and the items are exposed directly as capabilities
2. a virtual thing - a 'thing' that doesn't exist in openHAB but will be exported as device and then you manually at OH2/OH1 items to that virtual thing to be exposed as capabilities.  This allows you to use OH1 items and to basically create NEEO devices that don't exist in openHAB or NEEO devices that may be a combination of items (from other things).

Not sure if I'm clear on this - but bottom line it sounds like I need to add more verbage to this section to explain that better?  agree?

tmrobert8 commented on addons/io/org.openhab.io.neeo/src/main/java/org/openhab/io/neeo/internal/discovery/package-info.java:
```diff
@@ -0,0 +1,14 @@
+/**
```
Darn - was hoping I could avoid slapping the attribute on every class by using package-info - but I'll change it around to avoid package-info classes.

tmrobert8 commented on addons/binding/org.openhab.binding.neeo/README.md:
```diff
@@ -0,0 +1,254 @@
+
+A NEEO Recipes is a sequence of steps that accomplish a single task (launching a Scenario or turning off a Scenario).  You can view/modify Recipes using the NEEO app.
+
+You can run a Scenario by sending ON to the Scenario status channel and end the Scenario by sending OFF to the Scenario status channel.  Likewise, you can send ON to any "launch" type Recipe status channel to start the Scenario and send ON to the "poweroff" type Recipe status channel to end the Scenario.  Sending OFF to any Recipe status channel does nothing.
```
I'm modeling exactly how NEEO modeled it and I don't really have an answer here.  I think it has something to do with their 'automatic wiring' that they do (where a scenario may include power management for other devices - so there is a status for the scenario itself and then you can have 'launch'/'poweroff' recipes for other devices within the scenario being executed).  Pure guess however..

tmrobert8 commented on addons/binding/org.openhab.binding.neeo/README.md:
```diff
@@ -0,0 +1,254 @@
+
+| Channel Type ID    | Read/Write | Item Type    | Description                                                                                |
+|--------------------|------------|--------------|--------------------------------------------------------------------------------------------|
+| activeScenarios    | R          | Trigger      | The name of the active Scenarios (JSON array of names.  Example: "['TV','DVR']")           |
```
@martinvw asked for both of these channels to be triggers in a prior review.  I believe that the argument was that both this (and currentStep) are simply event notification from the brain and being an event notification - they should be trigger channels.

I fully agree with the 'currentStep' being a trigger because it really is just an event notification.  There is no state associated with it and it's even local to whatever process started the scenario (so if the remote starts the scenario - it get's the step notifications only, if openHAB starts the scenario, it get's the step notifications only).

I'm on the fence whether activeScenarios should be a trigger or item channel.  It's kinda a hybrid in that it's simply a notification from the brain - but depending on the underlying device, it may have state.  Example: starting a LIGHT, the scenario popup then goes away because lights don't have any state (so it's a notification).  However, start a TV - the activeScenario will include the TV until you are done watching it - so it has state.  What's your thoughts?  I'm leaning towards item channel for this one again just because it covers both situations.  This predated StringListType but I'll switch over to it with these changes.

triller-telekom replied:
I see, so in case of the currentStep you only get a notification and you can query which step you are in, right? In this case I agree with it being a Trigger channel.

And if the activeScenarios cannot be queried in all cases either, it should stay a Trigger channel. However I would like it to have a user friendly payload, i.e. a StringListType which can be rendered by UIs instead of a pure JSON array. WDYT?

tmrobert8 replied:
Correct in both cases - I'll leave it as a trigger but change it to a StringListType.  

FYI - Off conversation topic but I'm also converting over to the new Channel*Builder(s) you committed as well (change will be in MetadataUtils.java) - so you'll see that as a change in the next commit.

tmrobert8 replied:
Actually - I think I see why I did a JSON string - trigger channel only accepts strings (so I can't give it a StringListType).  Maybe you see something I don't?

triller-telekom replied:
Ok, let's take a step back: What is the purpose of this channel? Do you want to show the scenarios which are active in a UI? In this case this channel is clearly a state channel.

However, if you cannot fetch this information, i.e. react on a `REFRESH` command to obtain the current value for this channel, I have a hard time of thinking how this can be a useful information, i.e. if its not up2date most of the time.

The other option would be that you use the pushed information from the brain about the active scenarios for another purpose internally in the binding. So whenever the brain announces its active scenarios and the user wants to start one of them you just return instead of trying to start it again.

triller-telekom replied:
I just looked up in the code that only `CallItem` supports the `StringListType`.

But did you see this part of my comment?

> However, if you cannot fetch this information, i.e. react on a REFRESH command to obtain the current value for this channel, I have a hard time of thinking how this can be a useful information, i.e. if its not up2date most of the time.

So I would rather vote for removing this channel.

triller-telekom replied:
How about updating this channel with the information instead?

https://github.com/openhab/openhab2-addons/pull/3191/files/4d36ce3d7e6efd945b9556b8b4400e52ba9ec558#diff-1d97edbca8166155b4d0a32ae9c36b83R213

tmrobert8 replied:
You can fetch this information (that's why I was on the fence on this being a channel) and I know people do use this information (however, only in an event type of way - again, on the fence!).  However, you make a fair point with the status - let me think through some situations and see if they would ever be different (on the face, they shouldn't be different).  If they aren't, I'd tend to agree with you that maybe the channel is not needed.

tmrobert8 commented on addons/io/org.openhab.io.neeo/web/index.html:
```diff
@@ -0,0 +1,637 @@
+    <head>
+        <meta charset="UTF-8">
+        <title>NEEO Transport</title>
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
```
Note sure what you mean by "They should be included as a library in this PR".  I'm assuming you mean put them in my web\js directory or are you talking something else?

triller-telekom replied:
Yes, right. It should be possible to use your binding without an internet connection. So please deliver the libraries locally with your code.

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/NeeoConstants.java:
```diff
@@ -0,0 +1,95 @@
+    public static final String BINDING_ID = "neeo";
+
+    /** Threadpool ID */
+    public static final String THREADPOOL_ID = "neeo.binding";
```
Actually its not called "binding" :) Its called "thingHandler", see https://github.com/eclipse/smarthome/blob/36a470bfceb32e1eee511142a298dd6608a0b85c/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L71-L75

So inside your handlers you should make use of the `scheduler` instance from the `BaseThingHandler`.

tmrobert8 commented on addons/binding/org.openhab.binding.neeo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,132 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- The brain bridge - will have things dynamically assigned to it. -->
+	<bridge-type id="brain">
```
BTW - this was an EXCELLENT exercise you had me do.  I discovered an issue that has been bugging me for awhile in how I was dynamically generating the macros - was a minor issue so I never put the time into it - but doing this I found the issue (and realized it may be more than a 'minor' issue in certain circumstances).

tmrobert8 commented:
@triller-telekom 
Just pushed the changes (remove the activeScenarios channel!) and there were some bug fixes that I implemented with this version as well (so you may see some unrelated changes).  Do NOT approve yet - there is some weird bug in the dynamic part that I'm tracking down but I though you can restart your review in the meantime.

Thanks,
Tim

tmrobert8 commented:
@triller-telekom 
Thanks for the reply - I'll update it soon (and I have the new discovery process working - so I'll be updating that as well).

As for your question about the dynamically generated room types.  The one part you might have overlooked is that the groups are different (and that's the one thing you can't do from a thing).  The groups do contain the same channels - but they are all named differently.  So I have a group called "Recipe for TV (launch)" with the name/type/enabled/status channels and then I have a group called "Recipe for Soundbar (launch") with the same 4 channels.  Same applies to scenarios - I create a scenario group for each device.  Because of that - each room type is unique compared to the others.

For the device however - I'm on the fence.  A prior version did the same (where each macro had a group of channels).  I've simplified that so there is only a single channel now (and grouped by 'macro') - so I could change that back to a simple thing type and have the thing dynamically generate the channels.  However, I'm a little worried that I might go back to a group of channels because the latest firmwares seem to be making macros complicated again.  If that's the case, I'd rather bite the bullet now and keep everything to make the future easier.  As I said, on the fence on the device front - what's your thoughts?

As for the other comments - will do those shortly...

tmrobert8 commented:
@triller-telekom 
Just pushed your changes pending the above discussion.  Please note this will NOT compile because it uses the 'thing' discovery process that is pending discussion - see the other thread for that

tmrobert8 commented:
@triller-telekom 
Think I'm on board with the device being non dynamic (except for the channels - which will dynamically be created in the handler).  I'll commit tomorrow..

triller-telekom commented:
Thank you for agreeing to remove the dynamic device thing type creation.

Regarding the room thing-type creation: Did you notice that `Thing`s **do not** have `ChannelGroup`s? A `ThingType` **does have groups, but merely for the purpose of referencing "a group" of channels multiple times in the configuration xmls.

At runtime however, the groups are not existent, there is no `Thing.getChannelGroups()` because all channels are directly connected to the `Thing`, there is not group in between.

The only feature you gain from using channelgroups in your case is that the IDs of the channels will be prefixed with the groupname id, so you have channel ids like "groupId#channelID", or in your case something like "TVRecipe#status".

This makes me doubt that you need the dynamic `ThingType` at all, because you can easily create such channel IDs in your dynamically created and to be added channels too.

I will comment on the discovery in https://github.com/eclipse/smarthome/issues/5582

tmrobert8 commented:
I agree with you but then what are the purposes of groups to begin with?  I though groups (in the thing type) allowed the paperui (or other uis) to properly group the items together (which the paperui does wonderfully).   Personally I'd love to get rid of the dynamic stuff but am afraid it makes the paperui a mess then

triller-telekom commented:
Actually I was not aware that PaperUI renders the channel groups. Thanks for pointing this out.

However we do not have the information at runtime on the `Thing`. We have an open issue for more than 2 years now: https://github.com/eclipse/smarthome/issues/1118 But no one seems to be really needing this feature.

Since Paper UI is also an administration UI and most users will only see the usage UIs like basicUI, habpanel, android app, etc. they wan't see this anyway.

I agree with you that its nice that Paper UI groups the channels, but if they are sorted by name and you have the same prefix for all channels that you want to group together, they should also be shown together, right? Also the "Control" tab of Paper UI does not show channel groups, does it?

The purpose for having those groups was that if you have 2 thing types which have the same n channels, you put them in a group and reference the group in both thing types. It was not meant for orchestrating channels in a hierarchy.

tmrobert8 commented:
hmm - very good point about the control (it does not).  I focus on the paperui because that's what most of the people from NEEO use (I've funneled alot of people from the NEEO forums over to installing openHAB - but they are strictly new comers and non-techs that don't need nor want to edit files [items, sitemaps, etc] - they just want a plug and play interface and the paperui is great for that).   Once they have it installed and running - the NEEO remote becomes their interface to openHAB instead of any of the UIs - so the paperui (being the administrative) is the only thing they see.

So now I'm in a quandry
1. I see your point about the control and understand your points about the groups - do I just kill the grouping and get rid of the dynamic (which makes everyone's like easier on the development side) type?
2. Or do I try to plunge into that bug note and do something about groups and revisit the neeo one?

I guess 1 is the better option because I want to get this merged and off my plate and I've now delayed it because of the thing discovery stuff and doing 2 would take forever (scope, negotiations with everyone, actually doing the work, etc).

What do you think? 

Note: if you could comment (or get someone else to comment) on the thing discovery - I'd love to get that approved/merged quickly so we can get this off our plates as well..

triller-telekom commented:
I also think 1. is the better option (at least for now). Negotiations always consume a lot of time, but they also help to make the code and framework better. I also would like to see your binding merged as I have spend quite some time in reviewing it so it adheres to the concepts for bindings of openHAB and ESH.

I have commented on the discovery issue, would love to see a PR from you on the smarthome projects, after you added my comments which I left there.

tmrobert8 commented:
Will do - one followup question on the discovery before I make changes and submit a PR.  

Kinda off topic but hope you'd know.  In eclipse, my working sets have gotten all screwed up somehow (had a BSOD and it must have corrupted something).  Now all projects are under the "Other Projects" working set.  I've tried to run the "Perform initial tasks" (which recreated the other working sets) but it didn't move the projects back to the correct working sets.  Happen to know how to fix the working sets or know where a list of what should be in what working set?  The openhab and infrastructure ones were easy to recreate - but I don't know what goes into the Runtime or ESH Extensions working sets (guessing bindings, io, ui go into the extension but am not sure).

Thanks,
Tim

triller-telekom commented:
Sorry I cannot help you with the working sets because I just ignore them. Like you I never knew which projects to add to which working set, so I decided to use the plain project list. Also if a new bundle was added I never knew to which set to add it... But they are just for your personal use anyway, so you can ignore them like I do :)

tmrobert8 commented:
Just pushed the removal of all the dynamic types and updated documentation.  Last up is the whole discovery stuff..

tmrobert8 commented:
@triller-telekom 
Just pushed the 'hue' discovery process

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/NeeoHandlerFactory.java:
```diff
@@ -43,6 +51,8 @@
     @NonNullByDefault({})
     private NetworkAddressService networkAddressService;
 
+    private final Map<ThingUID, @Nullable ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
```
Did you put the `@Nullable` for the values here on purpose because you are putting values with `null` in here? or did you do that because a `get()` on a map may return `null`. I am assuming you did it because of the later and you should not do that because this would be part of external null annotations, see https://github.com/eclipse/smarthome/issues/5908

tmrobert8 replied:
lol - I did that because I copied it directly from the HUE thing factory and didn't give it a second of thought.  I removed it from mine  ;)

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/NeeoDeviceDiscoveryService.java:
```diff
@@ -52,40 +46,45 @@
-    @Override
-    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
-        return DISCOVERABLE_THING_TYPES_UIDS;
+    /** The timeout for searching the room */
```
What is the unit of this search time? seconds, minutes?

triller-telekom commented on addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/NeeoRoomDiscoveryService.java:
```diff
@@ -9,83 +9,82 @@
-    @Override
-    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
-        return DISCOVERABLE_THING_TYPES_UIDS;
+    /** The timeout for searching the brain */
```
see above, please mention the unit

triller-telekom commented:
Thank you for all the changes!

Can you please rebase your commits onto the current master to resolve the conflicts in the pom and feature file? They result from a merge of another binding in the meanwhile...

Also if you rebase anyway, could you please do an interactive rebase, i.e. `git rebase -i` so you can edit your commit messages? because the checker says that you are missing a correct sign-off in 9 commit.

Afterwards we are good to merge from my side :)

tmrobert8 commented:
@triller-telekom 
Yep - noticed those on the last commit.  I just commit your changes - if you would, do NOT approve this just yet.  Since we made so many changes, I need to run through the whole binding/transport to make sure there are no side effect bugs introduced.  Should give the thumbs up later today if I don't find anything..

tmrobert8 commented:
@triller-telekom 
Think we are good with the latest commit - looks like jenkins is complaining about the modbus - doesn't look like anything related to this build (but I don't think it got to the neeo binding because of it)

kaikreuzer commented on addons/binding/org.openhab.binding.neeo/README.md:
```diff
@@ -0,0 +1,289 @@
+# Neeo Binding
+
+This binding will help openHAB control a NEEO smart remote (specifically the NEEO Brain).
```
"will help" sounds a bit weird. How about simply "lets"

kaikreuzer commented on addons/binding/org.openhab.binding.neeo/README.md:
```diff
@@ -0,0 +1,289 @@
+This binding will help openHAB control a NEEO smart remote (specifically the NEEO Brain).
+NEEO is a smart home solution that includes an IP based remote.
+More information can be found at [NEEO](neeo.com) or in the forums at [NEEO Planet](https://planet.neeo.com).
+**This binding was not developed by NEEO** - so please don't ask questions on the NEEO forums.
```
This is the case for 99% of all bindings. Unless there is any very strong reason why we have to state this here, please remove this line.

tmrobert8 replied:
This was a condition from NEEO itself when I started it all - they wanted me to make it quite clear that they are not the source of the binding/io and will not support it.

kaikreuzer commented on addons/binding/org.openhab.binding.neeo/README.md:
```diff
@@ -0,0 +1,289 @@
+# Neeo Binding
```
Is it NEEO or Neeo? I assume the former.

kaikreuzer commented on addons/io/org.openhab.io.neeo/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,61 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Neeo Transport
```
This is actually not a "transport" bundle, but an I/O add-on. We don't have a fixed naming scheme for those, but e.g. Homekit and ImperiHome call them "Integration", so I'd suggest to go for the same: "NEEO Integration"

kaikreuzer commented on addons/io/org.openhab.io.neeo/README.md:
```diff
@@ -0,0 +1,411 @@
+# openHAB NEEO Transport
```
-> NEEO Integration

kaikreuzer commented on addons/io/org.openhab.io.neeo/README.md:
```diff
@@ -0,0 +1,411 @@
+# openHAB NEEO Transport
+
+This transport will allow openHAB things/items to be exposed to the NEEO Brain and allow the NEEO smart remote to control them.
```
-> This add-on allows ....

kaikreuzer commented on addons/io/org.openhab.io.neeo/README.md:
```diff
@@ -0,0 +1,411 @@
+This transport will allow openHAB things/items to be exposed to the NEEO Brain and allow the NEEO smart remote to control them.
+NEEO is a smart home solution that includes an IP based remote.
+More information can be found at [NEEO](neeo.com) or in the forums at [NEEO Planet](planet.neeo.com).
+**This transport was not developed by NEEO** - so please don't ask questions on the NEEO forums.
```
please remove this line

kaikreuzer commented on addons/io/org.openhab.io.neeo/README.md:
```diff
@@ -0,0 +1,411 @@
+More information can be found at [NEEO](neeo.com) or in the forums at [NEEO Planet](planet.neeo.com).
+**This transport was not developed by NEEO** - so please don't ask questions on the NEEO forums.
+
+The openHAB NEEO transport allows mapping of openHAB things/item to a NEEO device/capabilities (see Mappings below).
```
transport -> integration
Same for all other occurences below

kaikreuzer commented on addons/io/org.openhab.io.neeo/cfg/neeo.cfg:
```diff
@@ -0,0 +1,31 @@
+############################## openHAB NEEO Transport #############################
```
Integration

kaikreuzer commented on addons/io/org.openhab.io.neeo/pom.xml:
```diff
@@ -0,0 +1,30 @@
+	<artifactId>org.openhab.io.neeo</artifactId>
+	<packaging>eclipse-plugin</packaging>
+
+	<name>openHAB NEEO Transport</name>
```
Integration

kaikreuzer commented on addons/io/org.openhab.io.neeo/pom.xml:
```diff
@@ -0,0 +1,30 @@
+
+	<name>openHAB NEEO Transport</name>
+
+	<dependencies>
```
Do you really need to declare these dependencies here? The package imports in the manifest should be enough forthe resolution through Tycho afaik.

kaikreuzer commented on addons/io/org.openhab.io.neeo/src/main/java/org/openhab/io/neeo/AbstractServlet.java:
```diff
@@ -0,0 +1,227 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.io.neeo;
```
Imho most of your classes should be moved in an "internal" package and not be exposed.

kaikreuzer commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -310,6 +310,11 @@
         <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.modbus/${project.version}</bundle>
     </feature>
 
+    <feature name="openhab-binding-neeo" description="NEEO Binding" version="${project.version}">
```
What about the org.openhab.io.neeo bundle? Shouldn't you also add another feature for it in the I/O section?

kaikreuzer commented on addons/io/org.openhab.io.neeo/about.html:
```diff
@@ -0,0 +1,32 @@
+    indicated below, the terms and conditions of the EPL still apply to any source code in the Content
+    and such source code may be obtained at <a href="http://www.openhab.org/">openhab.org</a>.
+</p>
+
```
You are lacking the 3rd party section here.
Your bundle includes a whole lot of fonts, JS libraries, etc, which you need to list here with their licenses and a reference to their sources/projects (see other add-ons as an example).

tmrobert8 commented:
@kaikreuzer 
Thanks for the comments - heading off on a week vacation and will address them when I get back.

tmrobert8 commented:
@kaikreuzer 
As it turns out - the changes were really easy.   Committed...

tmrobert8 commented:
@kaikreuzer 
When I took the dependencies out of the pom.xml - the build is failing because it can't find them (even though they are in the manifest.mf).  If you happen to know why - I'll update it or I can revert.  Please advice..

kaikreuzer commented:
@tmrobert8 Thanks for your quick update. You are right, the dependencies are required for the build (they are not made available by default, because bindings should not depend on those, only I/O addons as in your case). I have just readded the dependencies and the build looks good - so let's merge 🎉 .

tmrobert8 commented:
Thanks @kaikreuzer and thanks @triller-telekom for the time and effort on this - I'll close the related issue

nitzpipp commented:
@tmrobert8 i have a problem with the neeo transport and kodi. when i try to use the ok-button on the remote, kodi is not reacting - up, down, right, left is working, but 'ok' is not. am i doing something wrong?
![select](https://user-images.githubusercontent.com/7882530/46317885-e9bcc000-c5d4-11e8-8cf0-96ab5c5f2cc2.png)
